### PR TITLE
[core] Add possibility of overriding paint properties inside format expression

### PR DESCRIPTION
**Pull request by alexshalamov**
This PR introduces the notion of _overridable_ properties for symbol layer and adds possibility to define `text-color` for `text-field` sections.

In case of symbol layer, style creator can define, for example, `text-color` paint property and text will be painted using defined color. Format section allows to override _default_ color on per-section basis, thus, `text-color` is _overridable_ paint property.

Whenever layout property `text-field` has overrides for paint properties, new `FormatOverrideExpression` is created. `FormatOverrideExpression`  expression, depending on the context, evaluates to value defined by paint property or to the value defined in `text-field` section.

In order to synchronize runtime styling with internal `FormatOverrideExpression`, extra check is added to `SymbolLayer::Impl::hasLayoutDifference` to identify whether constant paint property is changed.

Snippet from the style definition using paint properties within format expression.

```json
"layout": {
"text-field": ["format",
                ["get", "first_line"], { "font-scale": 1.2, "text-color": "red"},
                "\n", {},
                ["get", "second_line"], {
                  "text-color": "yellow",
                  "font-scale": 0.8,
                  "text-font": ["literal", [ "DIN Offc Pro Bold", "Arial Unicode MS Bold" ]]
                 }
  ],
  "text-font": [ "DIN Offc Pro Italic", "Arial Unicode MS Regular" ],
  "text-padding": 4,
  "text-max-width": 7,
  "text-size": 20
},
"paint": {
  "text-color": "white",
  "text-halo-color": "hsl(35, 8%, 85%)",
  "text-halo-width": 0.5,
  "text-halo-blur": 0
}
```
## Remaining work
 - [ ] Add platform runtime style bindings
 - [ ] Implement gl-js counterpart and add render tests

## Launch Checklist
* [x]  briefly describe the changes in this PR
* [x]  write unit tests for new functionality
* [ ]  document any changes to public APIs
* [x]  post [benchmark scores](https://github.com/mapbox/mapbox-gl-native/pull/13904#issuecomment-469164192)

Fixes: #13813

### CODE REVIEW

**Comment by 1ec5** [comment]
This design would be much more compatible with the NSAttributedString representation proposed in #14062 than what’s proposed in #13904. With #13904, the NSExpression syntax could only be a literal translation of the JSON syntax, which means it’ll be a bit convoluted compared to what developers are used to for text formatting. It also seems more appropriate to keep each expression self-contained than to rely on cross-references to portions of expressions set on other layer properties. It isn’t unheard of for developers, for instance, to set one layer’s `text-field` to the `text-field` of another layer.

**Comment by kkaefer** [comment]
Agree with @1ec5, I think this syntax is much more readable compared to https://github.com/mapbox/mapbox-gl-native/pull/13904. We're losing the ability of "CSS classes" (where you could reuse a style), but having a self-contained expression is probably worth it because it makes it easier to author styles by hand, as well as build an interface for it.

We've discussed before whether we want to remove the paint/layout property distinction from the stylesheet, and instead disallow the `zoom` expression for the existing layout properties. This is essentially what Studio also exposes in its UI (i.e. as a UI user you don't know about the distinction between `paint` and `layout` properties, so allowing "paint" properties as part of a layout property doesn't seem too much of a weakening of our conceptual distinction here.

**Comment by LukasPaczos** [comment]
> ... as a UI user you don't know about the distinction between paint and layout properties ...

This is also true for the Android public APIs and that's why I agree with @1ec5 and @kkaefer that making the syntax self-contained will be much easier to understand for end users. Also, becuase the `format` expression's DSL is already prepared to accept an arbitrary number of options, this approach will be easier to execute implementation-wise.

File: include/mbgl/style/expression/expression.hpp [hunk 1]
```diff
@@ -25,18 +25,26 @@ class EvaluationError {
 
 class EvaluationContext {
 public:
-    EvaluationContext(float zoom_) : zoom(zoom_), feature(nullptr) {}
-    EvaluationContext(GeometryTileFeature const * feature_) : zoom(optional<float>()), feature(feature_) {}
+    EvaluationContext() = default;
+    explicit EvaluationContext(float zoom_) : zoom(zoom_) {}
+    EvaluationContext(GeometryTileFeature const * feature_) : feature(feature_) {}
```
Comment (pozdnyakov) [inline include/mbgl/style/expression/expression.hpp]: nit: also explicit?

**Comment by samanpwbb** [comment]
From the Studio team's point of view, this spec proposal will work well 👍 . In general, the hard spec changes for studio are those that either introduce expressions with new argument signatures, or when interdependencies across properties get introduced. I like that this feels like a natural extension of format. Here's a quick diagram of how the proposal would work in a tool like Studio:

override-sketch

File: include/mbgl/style/expression/expression.hpp (hunk 1 above)
Comment (alexshalamov) [reply include/mbgl/style/expression/expression.hpp]: :+1: kiitos

**Comment by alexshalamov** [comment]
@LukasPaczos @julianrex would you be able to add bindings for Android / iOS?

**Comment by LukasPaczos** [comment]
Sure @alexshalamov, let's merge this one and track the platform tailwork in https://github.com/mapbox/mapbox-gl-native/issues/14101 and https://github.com/mapbox/mapbox-gl-native/issues/14102.

**Pull request merged**